nucypher/policy/policies.py

```diff
@@ -165,30 +165,100 @@ def payload(self):
 
-        super().__init__(deferredList, *args, **kwargs)
-        self._block_until_this_many_are_complete = int(len(deferredList) * self.percent_to_complete_before_release / 100)
+    def __init__(self,
+                 f,    # TODO: typing.Protocol
+                 nodes,
+                 network_middleware,
+                 percent_to_complete_before_release=5,
+                 note=None,
+                 threadpool_size=120,
+                 *args,
+                 **kwargs):
+        self.f = f
+        self.nodes = nodes
+        self.network_middleware = network_middleware
+        self.args = args
+        self.kwargs = kwargs
+
+        self.completed = {}
+        self.failed = {}
 
-    def _cbDeferred(self, *args, **kwargs):
-        result = super()._cbDeferred(*args, **kwargs)
+        self.percent_to_complete_before_release = percent_to_complete_before_release
+        self._partial_queue = Queue()
+        self._completion_queue = Queue()
+        self._block_until_this_many_are_complete = math.ceil(
+            len(nodes) * self.percent_to_complete_before_release / 100)
+        self.released = False
+        self.when_complete = Deferred()  # TODO: Allow cancelling via KB Interrupt or some other way?
+
+        if note is None:
+            tp_name=f"{f} to {len(nodes)} nodes"
+        else:
+            tp_name = f"{note}: {f} to {len(nodes)} nodes"
 
-        if self.finishedCount == self._block_until_this_many_are_complete:
-            self._policy_locking_queue.put(None)  # TODO: It'd be rad to return a list of nodes here who were contacted, but it's non-trivial.
-        return result
+        self._threadpool = ThreadPool(minthreads=threadpool_size, maxthreads=threadpool_size, name=tp_name)
 
     def block_for_a_little_while(self):
         """
         https://www.youtube.com/watch?v=OkSLswPSq2o
         """
-        _ = self._policy_locking_queue.get()  # Interesting opportuntiy to pass some data, like the list of contacted nodes above.
-        self.log.debug(f"{self.finishedCount} nodes were contacted while blocking for a little while.")
-        return
+        if len(self.completed) < self._block_until_this_many_are_complete:
+            _ = self._partial_queue.get()  # Interesting opportuntiy to pass some data, like the list of contacted nodes above.
+            self.log.debug(f"{len(self.completed)} nodes were contacted while blocking for a little while.")
+
+    def block_until_complete(self):
+        if self.total_disposed() < len(self.nodes):
+            _ = self._completion_queue.get()  # Interesting opportuntiy to pass some data, like the list of contacted nodes above.
+
+    def _handle_success(self, response, node):
+        if response.status_code == 202:
+            self.completed[node] = response
+        else:
+            assert False  # TODO: What happens if this is a 300 or 400 level response?
+
+        if len(self.completed) == self._block_until_this_many_are_complete:
```

Comment:
Because of concurrency should this be the following for safety...
```suggestion
        if len(self.completed) >= self._block_until_this_many_are_complete:
```

Reply:
Hmm, no that's no good either, as it will put additional values through the queue, emit duplicate log emissions, etc.  We need to use the `_released` flag here.  Wanna open an issue?

Reply:
Filed https://github.com/nucypher/nucypher/issues/2190.